### DISCUSSION TITLE
react-table row onClick

### DISCUSSION THREAD

**Discussion by sonalk215** [TanStack/table#2295] [General] [answered]
I am using react-table latest version. I want to call a function on click of a row, and the function should be passed the data of the row clicked. Any pointers would really help

**Comment by deadcoder0904** [comment, marked as answer]
You can use `onClick` on `tr`. Here's an example (search for `onClick`):

```js
import * as React from "react";
import { useTable } from "react-table";

type Data = {
  actor: string;
  movie: string;
}

const borderStyle = {
  border: "1px dashed navy"
};

export default function App() {
  const origData = [
    {
      actor: "Johnny Depp",
      movies: [
        {
          name: "Pirates of the Carribean 1"
        },
        {
          name: "Pirates of the Carribean 2"
        },
        {
          name: "Pirates of the Carribean 3"
        },
        {
          name: "Pirates of the Carribean 4"
        }
      ]
    }
  ];
  const newData: Array<Data> = [];
  origData.forEach(actorObj => {
    actorObj.movies.forEach(movie => {
      newData.push({
        actor: actorObj.actor,
        movie: movie.name
      });
    });
  });
  const data = React.useMemo(() => newData, []);
  const columns = React.useMemo(
    () => [
      {
        Header: "Actor",
        accessor: "actor"
      },
      {
        Header: "Movies",
        accessor: "movie"
      }
    ],
    []
  );
  const {
    getTableProps,
    getTableBodyProps,
    headerGroups,
    rows,
    prepareRow
  } = useTable({ columns, data });
  return (
    <table {...getTableProps()}>
      <thead>
        {headerGroups.map(headerGroup => (
          <tr {...headerGroup.getHeaderGroupProps()}>
            {headerGroup.headers.map(column => (
              <th {...column.getHeaderProps()} style={borderStyle}>
                {column.render("Header")}
              </th>
            ))}
          </tr>
        ))}
      </thead>
      <tbody {...getTableBodyProps()}>
        {rows.map((row, i) => {
          prepareRow(row);
          return (
            <tr {...row.getRowProps()} onClick={() => console.log(row.original)}>
              {row.cells.map((cell, j) => {
                return (
                  <td
                    rowSpan={cell.rowSpan}
                    {...cell.getCellProps()}
                    style={borderStyle}
                  >
                    {cell.render("Cell")}
                  </td>
                );
              })}
            </tr>
          );
        })}
      </tbody>
    </table>
  );
}
```

You can alternatively paste this in [Codesandbox](https://codesandbox.io) & try clicking the rows to see it work :)

**Reply by jaseeyas** [reply]
can i get that clicked row value outside of table ? Pls share the code for the same.

**Reply by sonalk215** [reply]
You can call a function(defined in another component) and pass that function as a prop

    onClick={()=>props.testRowClick(row)

Hope that answers your question

**Reply by Niroznak** [reply]
thanks
looked a long time for something like this

**Reply by MichaelGreco** [reply]
You are a literal life saver!!! I have been looking to the solution to this for hours!!!

**Reply by efrenmarin45** [reply]
Really wish React-Table would have more comprehensive docs. I am struggling with the same issue. I have an onClick event on  <tr> and can console log the info for row.original. This is great since it shows me that wherever I click I can get that row's data. My issue is I can't pull and save that data anywhere else. Everywhere I try to call 'row' it comes as undefined. After reading through the forums, it seems like 'row' is only accessible through getRowProps. I have a Modal component that I want to pass this data to. 

I see an answer to call a function and pass props but for some reason this does not work for me. For context, my table is the parent and the Modal is the child. Any suggestions?

**Reply by sayinmehmet47** [reply]
@efrenmarin45  I solve it by saving the row.original to state then, sending this this state via redux to other components is the easiest way I think, Lets take a look at that 
```javascript
import React, { useEffect, useMemo, useState } from "react";
import { useDispatch, useSelector } from "react-redux";
import { useTable, useSortBy, usePagination } from "react-table";
import { updateFromCurrency } from "../store/Actions/currencyActions";
import { BuyingModal } from "./BuyingModal";
import { SellingModal } from "./SellingModal";

export const PortfolioTable = () => {
  const portfolio = useSelector((state) => state.portfolioData);
  const dispatch = useDispatch();
  const [fromCurrency, setFromCurrency] = useState("");
  const data = useMemo(() => [...portfolio], [portfolio]);

  useEffect(() => {
    dispatch(updateFromCurrency(fromCurrency));
  }, [fromCurrency]);

  const columns = useMemo(
    () => [
      {
        Header: "Acronym",
        accessor: "acronym", // accessor is the "key" in the data
      },
      {
        Header: "Name",
        accessor: "name",
      },
      {
        Header: "Amount",
        accessor: (d) => d.totalAsset.toFixed(3),
      },
      {
        width: 300,
        Header: "Trade",
        Cell: ({ row }) => (
          <div className="d-flex justify-content-center">
            <BuyingModal
              buttonLabel="Buy"
              selected={[row.original.acronym, row.original.name]}
            />
            <SellingModal buttonLabel="Sell" />
          </div>
        ),
      },
    ],
    []
  );

  const {
    getTableProps,
    getTableBodyProps,
    headerGroups,
    nextPage,
    previousPage,
    prepareRow,
    page,
    canPreviousPage,
    canNextPage,
  } = useTable({ columns, data }, useSortBy, usePagination);
  return (
    <div className="d-flex flex-column mb-5 mt-5 mx-5">
      <table {...getTableProps()} style={{ borderRadius: "15px" }}>
        <thead>
          {headerGroups.map((headerGroup) => (
            <tr {...headerGroup.getHeaderGroupProps()}>
              {headerGroup.headers.map((column) => (
                <th
                  {...column.getHeaderProps(column.getSortByToggleProps())}
                  className="border"
                  style={{
                    background: "#f3f4f6",
                    color: "black",
                    paddingLeft: "12px",
                  }}
                >
                  {column.render("Header")}
                  <span>
                    {column.isSorted ? (column.isSortedDesc ? "🔽" : "🔼") : ""}
                  </span>
                </th>
              ))}
            </tr>
          ))}
        </thead>
        <tbody {...getTableBodyProps()}>
          {page.map((row) => {
            prepareRow(row);
            return (
              <tr
                {...row.getRowProps()}
                onClick={() => setFromCurrency(row.original.acronym)}
              >
                {row.cells.map((cell) => {
                  return (
                    <td
                      {...cell.getCellProps()}
                      className="border py-2 "
                      style={{
                        textDecorationLine: "none",
                        paddingLeft: "7px",
                      }}
                    >
                      {cell.render("Cell")}
                    </td>
                  );
                })}
              </tr>
            );
          })}
        </tbody>
      </table>
    </div>
  );
};

```

**Reply by shehmirkz** [reply]
**It works**

**Reply by SethuSS** [reply]
thank you man it works

**Reply by abdulsamadayoade** [reply]
thanks man, it works

**Reply by sultanov2103** [reply]
а что напишим на store. Вы с помощью useSelector что  то взяли там

**Reply by Nik-Ark** [reply]
Will it be accessible to people with visual impairments ?

**Reply by adarshSrivastava01** [reply]
what if I have a generic component and when I call this <App /> component, I want to pass the action as <App onRowClick={() => console.log(row.id)} />, is it possible at all

**Comment by jaseeyas** [comment]
Thanks a lot for the reply. Also one more thing, how to highlight the  selected row on click? I need to style the row as per click. Pls help out.

**Reply by sonalk215** [reply]
You can set a class to each row 

        <tr
            {...row.getRowProps()}
            className={classes['table-body-row']}
            onClick={()=>props.testRowClick(row)}
        >

and in css

        .table-body-row:hover {
            background-color: #e6e6ff;
        }


where :hover is a css pseudoclass

Hope this helps

**Reply by vikrant-gembrill** [reply]
what will happen if I directly call the method defined in the parent component? also i am construction the columns in the parent component, and passing the columns and data as props to child component which renders the react-table.  i tried this and this is working but i am not getting the state properly? is this causing some kind of closure issues?

**Comment by jaseeyas** [comment]
but i cant get it on click..only on hovering its effects comes :(

Thanks and Very best regards

Yasirali

Abudhabi Autonomous Systems Investments

United Arab Emirates
Mobile: +971 (50) 9721325
E-mail: yasir.veetil@adasi.ae <YKURIKKALA@etihad.ae>
http://www.adasi.ae/ <http://www.etihad.ae/>

P.O. Box 109667, Abu Dhabi,

**Reply by deadcoder0904** [reply]
then add a class on click & style it with css

**Comment by jaseeyas** [comment]
As i am new, could you please provide a sample code. I didnt get how to add
a class on click. Any help appreciated :)

Thanks and Very best regards

Yasirali

Abudhabi Autonomous Systems Investments

United Arab Emirates
Mobile: +971 (50) 9721325
E-mail: yasir.veetil@adasi.ae <YKURIKKALA@etihad.ae>
http://www.adasi.ae/ <http://www.etihad.ae/>

P.O. Box 109667, Abu Dhabi,

**Comment by jaseeyas** [comment]
reminder pls

Thanks and Very best regards

Yasirali

Abudhabi Autonomous Systems Investments

United Arab Emirates
Mobile: +971 (50) 9721325
E-mail: yasir.veetil@adasi.ae <YKURIKKALA@etihad.ae>
http://www.adasi.ae/ <http://www.etihad.ae/>

P.O. Box 109667, Abu Dhabi,

**Reply by sonalk215** [reply]
Hi

try something like below

https://stackoverflow.com/questions/42630473/react-toggle-class-onclick

hope this helps...

**Comment by isomorpheric** [comment]
Does anyone know how to have that same onClick()  but only on a button in the last column of the row. I don't want users to click the entire row, just the "Edit" button for the row. If the 'row' is only accessed by 'getRowProps', how can that specificil cell access the row data?

**Reply by davewasthere** [reply]
For that you'd want to define a Cell in your columns definition.

e.g.

`Cell: ({value, column, row}) => { return( <div onClick={ () =>  doSomethingWithRow(row.original) }>{value}</div> ) }`

**Reply by isomorpheric** [reply]
Appreciate it, thanks!

**Reply by nani1008** [reply]
Hello @ecam900, How can I get the props from the Cell and display it on the jsx? Any help! Im stuck with this! Thank you!

**Comment by mohammedameen028** [comment]
I want to add a new row(at the end) with the click of the "Add Row" button in react table created. Can someone please me?

**Reply by MichaelGreco** [reply]
When you click an add row button that you make, have it open a Modal that collects the info that is needed for that row. Add it to the useState variable that you are using for the table rows and it should automatically update

**Comment by merski007** [comment]
Anyone else notice a delay when using the toggleRowSelected function inside the onClick event? For example, the state of selectedFlatRows does not update until after the onClick event is complete. On first click, the isSelected prop is false, even though selectedFlatRows will display data outside of the event. And then on second click, the prop is true, even though selectedFlatRows is empty.
Code:
![image](https://user-images.githubusercontent.com/31394435/168896126-d4e8b33a-31bf-4cea-96b8-92e330cee149.png)


Results:
![image](https://user-images.githubusercontent.com/31394435/168896006-666a5438-927d-493b-97a4-a506769e6938.png)
